solr/core/src/test/org/apache/solr/search/neural/KnnQParserTest.java

```diff
@@ -453,52 +455,397 @@ public void knnQueryUsedInFilter_shouldFilterResultsBeforeTheQueryExecution() {
   @Test
   public void knnQueryUsedInFilters_shouldFilterResultsBeforeTheQueryExecution() {
     String vectorToSearch = "[1.0, 2.0, 3.0, 4.0]";
+
+    // topK=4 -> 1,4,2,10
     assertQ(
         req(
             CommonParams.Q,
             "id:(3 4 9 2)",
             "fq",
             "{!knn f=vector topK=4}" + vectorToSearch,
             "fq",
-            "id:(4 20)",
+            "id:(4 20 9)",
             "fl",
             "id"),
         "//result[@numFound='1']",
         "//result/doc[1]/str[@name='id'][.='4']");
-  }
-
-  @Test
-  public void knnQueryWithFilterQuery_shouldPerformKnnSearchInPreFilteredResults() {
-    String vectorToSearch = "[1.0, 2.0, 3.0, 4.0]";
 
+    // topK=4 w/localparam preFilter -> 1,4,7,9
     assertQ(
         req(
             CommonParams.Q,
-            "{!knn f=vector topK=10}" + vectorToSearch,
+            "id:(3 4 9 2)",
+            "fq",
+            "{!knn f=vector topK=4 preFilter='id:(1 4 7 8 9)'}" + vectorToSearch,
             "fq",
-            "id:(1 2 7 20)",
+            "id:(4 20 9)",
             "fl",
             "id"),
+        "//result[@numFound='2']",
+        "//result/doc[1]/str[@name='id'][.='4']",
+        "//result/doc[2]/str[@name='id'][.='9']");
+
+    for (String fq :
+        Arrays.asList(
+            "{!knn f=vector topK=5 includeTags=xxx}" + vectorToSearch,
+            "{!knn f=vector topK=5 excludeTags=xxx}" + vectorToSearch)) {
+      assertQEx(
```

Comment:
Also in this case I would prefer to see these as separate test cases, that can run independently